Add IP address field type.

mgetka opened the pull request:
This pull request adds IPv4/IPv6 fields serialization and deserialization w/ tests included. To avoid confusion I am excluding the possibility to specify IP addresses as an integer or byte values.

mgetka committed:
Add IP address field type.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1629,6 +1631,42 @@ def __init__(self, *args, **kwargs):
+                # confusion.
+                raise TypeError(
+                    "Only dot-decimal and hexadecimal groups notations are supported. "
+                    "Got %s."
```
I can't find a source to back this up but I think we have a consensus on not returning input values in error messages.

mgetka replied:
Fixed.

mgetka committed:
Skip input value logging in IP field validation

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1629,6 +1631,41 @@ def __init__(self, *args, **kwargs):
+        """Format the value or raise a :exc:`ValidationError` if an error occurs."""
+        if value is None:
+            return None
+        if isinstance(value, (ipaddress.IPv4Address, ipaddress.IPv6Address)):
```
Currently, marshmallow does not accept deseralized form when deserializing. This is discussed in #1415.

mgetka replied:
Dropping lines 1645-1646 will actually have no effect on the interface level since `ipaddress.ip_address` function accepts `ipaddress.IPv4Address` and `ipaddress.IPv6Address` instances, and returns their copies.

Are you suggesting to add explicit type checking and accept only `str` values?

lafrech replied:
I see no need to explicitly exclude deserialized form. But that could be the reason I'm in favour of #1415. I didn't check but I'd bet other fields already accept deserialized data.

In any case, I see no need for this check. As you said, the behaviour will stay the same.

mgetka replied:
In that case, the issue is resolved via 985f2ba

lafrech reviewed (commented):
Should we add a parameter to only accept IPv4 or IPv6 rather than accepting any?

Should we also provide `IPv4` and `IPv6` fields as a shortcut?

Also, I don't see the point of the `_serialize` method. Apparently you copied it from an already existing field (`UUID`) so it's nothing new. But this seems a bit dodgy to me. Let's discuss this in #1489.

mgetka committed:
Omit redundant serialization step for IP field

mgetka commented:
> Also, I don't see the point of the _serialize method. Apparently you copied it from an already existing field (UUID) so it's nothing new. But this seems a bit dodgy to me.

That is correct, I've overridden the `_serialize` based on insights collected from `UUID` field implementation. After inspecting the implementation of `String` base class, I can no longer see any purpose of `_serialize` method.

mgetka committed:
Drop deserialized value handling in IP field validation

mgetka committed:
Allow for IPv6 exploded form serialization

mgetka committed:
IP v4/v6 specific fields added

mgetka committed:
Merge branch 'dev' into ip-address-field

mgetka committed:
Expose IP v4/v6 specific fields for wildcard imports

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1634,6 +1638,91 @@ def __init__(self, *args, **kwargs):
+            return None
+        try:
+            if isinstance(value, (int, bytes)):
+                raise TypeError("Only dot-decimal notation is supported.")
```
What about hexadecimal groups as above in generic IP field?

mgetka replied:
IPv6 error message mistake. Dot decimal notation is exclusive for IPv4 and hexadecimal groups are appropriate for IPv6. Fixed.

mgetka committed:
Fix exception message on invalid IPv6 field value

mgetka commented:
Also please note that _exploded_ notation is specific for IPv6 addresses, but the argument is also present in IPv4 constructor. This feature is sourced from [`ipaddress` python standard library](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.exploded), where exploded notation can be applied to both IPv4 and IPv6 for consistency.

mgetka committed:
Merge branch 'dev' into ip-address-field

sloria commented:
Would it make sense to implement these as validators? That would allow combining with other validators.

mgetka committed:
IP addresses fields inherits from Field base class

mgetka committed:
Move _validated method functionalities of IP fields  into _deserialize

mgetka commented:
I have to admit that maybe I was too inspired by the implementation of the `UUID` field, and implemented the functionality without more complete understanding of fields module structure. As I can see, fields inheriting from `String`, are in general deserialized from strings to strings with additional input value format validation. We can see it in the `Url` and `Email` fields cases. Exception from this rule is the `UUID` field, which deserializes from string into UUID object representation. Therefore, it is not clear for me why `UUID` inherits from `String` and not `Field`, like in the case of `DateTime` or `TimeDelta` which also deserializes to object representation.

`ipaddress` library utilities provides with IP address object representations with validation occurring on object construction. In my opinion creation of the object representation for the purpose of validation, and then discarding it would be an overkill. Also I can see clear contract between _validators_ based implementation of `Url` and `Email` fields and  `String` base class features. But IP addresses are not the case - they are more like `DateTime` or `TimeDelta` fields. Therefore, IP field now inherits from `Field` and not `String`.

Also, I can see, that `_validated` method is utilized for chained validation in inheriting fields classes.
For IP fields it is not the case, since IPv4 and IPv6 Fields implements its own independent validation procedures. I have moved validation/object creation into `_deserialize` methods.

Returning to the case of the UUID field. Maybe it should be discussed whether it should inherit from `String` or rather `Field` to get a more transparent structure of the module?

I am also wondering whether IP specific validators are welcome? Based on `ipaddress` functionalities it is straightforward to implement validators like: `PrivateIPAddress`, `GlobalIPAddress` or one that checks whether IP address belongs to specified subnetwork.

mgetka committed:
Merge branch 'dev' into ip-address-field

lafrech commented on tests/test_deserialization.py:
```diff
@@ -841,6 +842,83 @@ def test_invalid_uuid_deserialization(self, in_value):
+        result = field.deserialize(ipv4)
+        assert isinstance(result, ipaddress.IPv4Address)
+        assert result == ipv4
+
```
Do we have to test this?

In #1415, we discuss accepting deserialized form as input to deserialization methods. For now it only works if the deserialization method happens to accept deserialized form, like this one does, but it is more a happy accident than a design choice.

I don't have a strong opinion about this test. But until we go on with #1415, I'm lukewarm to guarantee this behaviour for a given field.

Are there other field for which we test this?

mgetka replied:
And again, such case is present in UUID field tests. For the IP fields, I have removed tests on unintended features like this.

lafrech reviewed (commented):
Looks good to me.

Just a question/comment about a test. I wouldn't mind another opinion here.

Your remarks about UUID field make sense. Reworking it could be done in another PR, and perhaps in marshmallow 4 if it is a breaking change.

IP specific validation welcome. Should those be field constructor arguments instead like exploded?

deckar01 commented on src/marshmallow/fields.py:
```diff
@@ -1616,6 +1620,79 @@ def __init__(self, *args, **kwargs):
+        if value is None:
+            return None
+        try:
+            if isinstance(value, (int, bytes)):
```
Fields that inherit from `String` don't have to store their deserialized value internally as a string. Because fields are both dumpers and loaders, it is better to think of `String` as `String -> Any -> String`. Subclassing `String` allows you to specialize `Any` to a more concrete type.

IMO explicitly checking for ints and bytes is unnecessary. If you inherit form `String`, `super()._deserialize()` will provide all the type validation that is needed. `ipaddress` takes care of validating the format.

mgetka replied:
`ipaddress` utilities are flexible in the terms of input ip addresses representations. Beside classic string representations it also accepts binary and integer representations. Explicit type checking was my arbitrary choice in order to narrow down the original `ipaddress` functionality. In my opinion deserialization of following JSON
```
{"ip": 1}
```
into this object representation
```
{"ip": IPv4Address('0.0.0.1')}
```
may be unexpected. But as I can see it leads to confusion on the implementation level, thus I have added constructor parameter determining whether decimal representations should be accepted. The parameter is `False` by default.

deckar01 replied:
Only passing strings to `ipaddress` seems reasonable. Excluding a subset of the types that `ipaddress` accepts seems fragile. If `ipaddress` starts accepting new types in the future this would let other types through. Allowing only strings would make the type checking explicit and resilient to upstream changes. The `String` field is a nice way to inherit this behavior declaratively.

mgetka replied:
Accepting only strings when deserializing IP fields is something that I can agree on. But achieving this via inheriting from `String` is something that is not so clear for me. From the sources and the discussion I can see two approaches on the `String` base class.

First approach is the one I described in previous comment, where `String` based fields deserialize into `str`, and have specific validator attached. Examples of such fields are `URL` and `Email` which are just string values with additional format constraints. If so, `UUID` filed nor `IP` field should inherit from `String`, since they are not based on a validator and doesn't serialize into string.

The second approach, is the one where `String` based fields are like `String -> Any -> String`. So, if I understand it correctly, all fields that serialize to strings, regardless of deserialized type, should inherit from `String`. Then, `UUID` and `IP` fields should inherit from `String`, as well as all `DateTime` and `Time` fields, which currently inherit from `Field`.

mgetka replied:
Since there is no feedback on this matter, I decided on an approach that better suits my understanding of the structure of the module.

deckar01 replied:
The only thing `String._deserialize()` would do differently is decode bytes into a UTF-8 string using `utils.ensure_text_type()`. That has been the default behavior for all of the string fields for a long time, but it's not clear what the use case was from the original commit.

@sloria Do you remember why byte strings are assumed to be UTF-8 encoded and decoded into strings? Should we do that for IP address strings also?

Original commit: https://github.com/marshmallow-code/marshmallow/commit/ff18e782be9d04557f027116b147f16bcecb9999

mgetka replied:
I suppose that it may be related to the fact that `Schema` objects can have `render_module` altered to some unusual one that does not perform data decoding? If it's so, it is correct to utilize `utils.ensure_text_type()` as well in IP fields, since `ipaddress` utilities will not treat `b"192.168.0.1"` the same way as `"192.168.0.1"`. I'll wait for additional feedback on that matter.

mgetka replied:
To be compatible with other field interfaces I've utilised `utils.ensure_text_type()` helper function. From the IP fields perspective it is acceptable - all expected behaviours are consistent with original approach. Numeric (ie. `int(1) -> 0.0.0.1`) and binary (ie. `b"\x00\x00\x00\x01" -> 0.0.0.1`) inputs will not be interpreted by `ip_address` library utilities. 

If the usage of `utils.ensure_text_type()` is obsolete or somehow stems from no longer applicable interface requirements, I belive it shpould be discussed in another issue, and appropriate patch should be applied on all fields utilizing the function.

mgetka committed:
Do not test unintended features of IP fields.

mgetka committed:
decimal representation deserialization in IP field

mgetka committed:
Revert "decimal representation deserialization in IP field"

This reverts commit 474e06ddf74e1260cc9890a2b62013e32c37210b.

mgetka committed:
IP fields accept only string encoded IP form

mgetka committed:
Merge branch 'dev' into ip-address-field

mgetka committed:
Ensure text type on IP fields deserialization

mgetka committed:
Partially revert "Ensure text type on IP fields deserialization"

This partially reverts commit f197cafeb012409a807e9357caeaa05888d4f1cc.

mgetka committed:
Merge branch 'dev' into ip-address-field

sloria commented:
@lafrech @deckar01 Can one of you do a final CR pass on this? Looks like it's close to--if not completely--ready to merge.

lafrech reviewed (commented):
LGTM. I'm just concerned about the code duplication. Could `_deserialize` be factorized, using class attributes for the object class, like what is done for the error message? (The typing would be a little less strict, but oh well.)

mgetka committed:
Merge branch dev' into ip-address-field

mgetka committed:
IP fields deserialization factorized

mgetka commented:
> LGTM. I'm just concerned about the code duplication. Could `_deserialize` be factorized, using class attributes for the object class, like what is done for the error message? (The typing would be a little less strict, but oh well.)

I've tried to define class variable typed as `typing.Callable[[str], typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]` which holds `ipaddress.ip_address` for generic IP field, and `ipaddress.IPv4Address` and `ipaddress.IPv6Address`  classes for IPv4 and IPv6 fields respectively, but there are issues with `ipaddress.ip_address` becoming a method. The issue can be suppressed by decorating the `ipaddress.ip_address` with `staticmethod` like
```
DESERIALIZATION_CALLABLE = staticmethod(ipaddress.ip_address)
```
but I'm not entirely comfortable with such construct. Also, mypy doesn't like it as well since there is unresolved issue [#3482](https://github.com/python/mypy/issues/3482) which is of relevance for class methods and static methods defined this way.

mgetka committed:
py35 compliant class attribute typing in IP fields

mgetka committed:
Fix duplicate entry in AUTHORS.rst

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1625,6 +1629,56 @@ def __init__(self, *args, **kwargs):
+
+    default_error_messages = {"invalid_ip": "Not a valid IP address."}
+
+    DESERIALIZATION_CLASS = None  # type: typing.Optional[typing.Type]
```
I'd write

    DESERIALIZATION_CLASS = ipaddress.ip_address

Then in `_deserialize`

            return self.DESERIALIZATION_CLASS(utils.ensure_text_type(value))

Unless there is a reason not to do that?

mgetka replied:
I've presented rationale for this in the comment to the pull request. Long story short, `ip_address` is a function and eagers to become a method when assigned to a class attribute.

lafrech replied:
Sorry, I did not understand your PR comment.

As you say, it works if we use `staticmethod`:

```py 
DESERIALIZATION_FUNC = staticmethod(ipaddress.ip_address)
```

mypy doesn't complain if I write


```py 
    DESERIALIZATION_FUNC = staticmethod(
        ipaddress.ip_address
    )  # type: typing.Union[staticmethod, ipaddress.IPv4Address, ipaddress.IPv6Address]
```

but I have little experience with Typing so I can't tell if it is the right way to go.

Type checking aside, I found the staticmethod alternative better but I'm starting to understand your rationale.

(Slight performance advantage for the staticmethod alternative because the check is not performed every load, but that's peanuts.)

lafrech reviewed (approved):
Sorry for the delay and thanks for your work on this.

lafrech commented:
@sloria, @deckar01, any objection to merging this?.

sloria commented:
@lafrech I didn't to a fine-toothed review of this, but it looked good last time I looked. I defer to you on merging this.

mgetka committed:
Merge branch 'dev' into ip-address-field

lafrech commented:
Merged and released.

Thanks again @mgetka.
